src/Backend/Haskell.idr

```diff
@@ -1,68 +1,127 @@
 module Backend.Haskell
 
-import Data.Vect
-import Control.Monad.State
-
-import Backend.Utils
 import Types
 import Typedefs
 
-%default partial
+import Backend
+import Backend.Utils
+
+import Text.PrettyPrint.WL
+import Control.Monad.State
+
+import Data.Vect
+
+%default total
 %access public export
 
-guardPar : String -> String
-guardPar str = if any isSpace $ unpack str then parens str else str
+||| The syntactic structure of Haskell types.
+data HsType : Type where -- TODO could be interesting to index this by e.g. used variable names?
```

Comment:
Nice! :+1:

Reply:
Are you thinking of the `TODO`? Because that turned out very messy, when I tried. OTOH the rest of my code hadn't stabilized then, so it might be easier now. Seems reasonable to take care of this as a separate issue once this monster has been merged though. :)